cloud/scope/powervs_cluster.go

```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+				if err != nil {
+					return false, fmt.Errorf("error validating security group rule remote: %v", err)
+				}
+			}
+		case "*vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolIcmp":
+			ogRule := ogRuleIntf.(*vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolIcmp)
+			code, _ := strconv.Atoi(*icmpCode)
+			iType, _ := strconv.Atoi(*icmpType)
+
+			if *ogRule.Direction == string(direction) && *ogRule.Protocol == protocol && *ogRule.Code == int64(code) && *ogRule.Type == int64(iType) {
+				ogRemote := ogRule.Remote.(*vpcv1.SecurityGroupRuleRemote)
+				match, err = s.validateVPCSecurityGroupRuleRemote(ogRemote, remote)
+				if err != nil {
+					return false, fmt.Errorf("error validating security group rule remote: %v", err)
+				}
+			}
+		}
+		if match {
+			break
+		}
+	}
+
+	return match, nil
+}
+
+// validateVPCSecurityGroupRules compares a specific security group rules spec with the existing security group's rules.
+func (s *PowerVSClusterScope) validateVPCSecurityGroupRules(originalSecurityGroup vpcv1.SecurityGroup, expectedSecurityGroup infrav1beta2.SecurityGroup) (bool, error) {
+	for _, expectedRule := range expectedSecurityGroup.Rules {
+		direction := expectedRule.Direction
+		var protocol string
+		var portMin, portMax int64
+
+		switch direction {
+		case infrav1beta2.SecurityGroupRuleDirectionInbound:
+			protocol = string(expectedRule.Source.Protocol)
+			portMin = int64(expectedRule.Source.PortRange.MinimumPort)
+			portMax = int64(expectedRule.Source.PortRange.MaximumPort)
+			icmpCode := expectedRule.Source.ICMPCode
+			icmpType := expectedRule.Source.ICMPType
+
+			for _, remote := range expectedRule.Source.Remotes {
+				match, err := s.validateSecurityGroupRule(originalSecurityGroup, direction, protocol, portMin, portMax, icmpCode, icmpType, remote)
+				if err != nil {
+					return false, fmt.Errorf("error validating security group rule: %v", err)
+				}
+				if !match {
+					return false, nil
+				}
+			}
+		case infrav1beta2.SecurityGroupRuleDirectionOutbound:
+			protocol = string(expectedRule.Destination.Protocol)
+			portMin = int64(expectedRule.Destination.PortRange.MinimumPort)
+			portMax = int64(expectedRule.Destination.PortRange.MaximumPort)
+			icmpCode := expectedRule.Destination.ICMPCode
+			icmpType := expectedRule.Destination.ICMPType
+
+			for _, remote := range expectedRule.Destination.Remotes {
+				match, err := s.validateSecurityGroupRule(originalSecurityGroup, direction, protocol, portMin, portMax, icmpCode, icmpType, remote)
+				if err != nil {
+					return false, fmt.Errorf("error validating security group rule: %v", err)
+				}
+				if !match {
+					return false, nil
+				}
+			}
+		}
+	}
+
+	return true, nil
+}
+
+// validateVPCSecurityGroup validates the security group and it's rules.
+func (s *PowerVSClusterScope) validateVPCSecurityGroup(securityGroup infrav1beta2.SecurityGroup) (*string, error) {
+	securityGroupDet, err := s.IBMVPCClient.GetSecurityGroupByName(*securityGroup.Name)
```

Comment:
May we can try to fetch by id as well, if id is already set in status? also we need to make sure name is not nil.